### DISCUSSION TITLE
关于Makefile的%与变量替换

### DISCUSSION THREAD

**Discussion by namezhenzhang** [physics-data/faq#113] [Q&A] [answered]
在如下代码中：
```Makefile
a_dep:=b.txt
b.txt :
	@echo "|$@ <- $^|"
%.txt : $(%_dep)
	@echo "|$@ <- $^|"
all : a.txt
```
第四行` : `右边的%没有被正确替换, 输出为:
```
|a.txt <- |
```

1. 这是由于什么原因导致的呢?

2. 如果我想实现这种变量替换的方式, 而不想将b.txt直接写上去, 该怎么做呢?

**Comment by jiegec** [comment]
1. Makefile 应该只支持文件名中 % 的替换：https://www.gnu.org/software/make/manual/html_node/Pattern-Rules.html
2. 可以参考 https://stackoverflow.com/questions/45156678/makefile-pattern-in-prerequisite-variable

**Reply by namezhenzhang** [reply]
感谢助教！

**Comment by heroxbd** [comment, marked as answer]
这个问题较复杂，需要使用二次替换，在 GNU Make 中叫 "secondary expansion"。

```
a_dep:=b.txt
b.txt :
	@echo "|$@ <- $^|"

.SECONDEXPANSION:
%.txt : $$(%_dep)
	@echo "|$@ <- $^|"
all : a.txt
```

先进行一次匹配变成

```
a_dep:=b.txt
a.txt : $(a_dep)
	@echo "|$@ <- $^|"
```

然后变量替换成

```
a.txt : b.txt
	@echo "|$@ <- $^|"
```

参见：https://www.gnu.org/software/make/manual/html_node/Secondary-Expansion.html

**Reply by namezhenzhang** [reply]
感想老师！
